src/golang/lib/engine/authenticate.go

```diff
@@ -43,12 +49,46 @@ func AuthenticateLambdaConfig(ctx context.Context, authConf auth.Config) error {
 		lambda_utils.SnowflakeConnectorType,
 	}
 
-	for _, functionType := range functionsToShip {
-		err := lambda_utils.CreateLambdaFunction(functionType, lambdaConf.RoleArn)
-		if err != nil {
-			return errors.Wrap(err, "Unable to Create Lambda Function")
-		}
+	errGroup, _ := errgroup.WithContext(ctx)
+
+	pullImageChannel := make(chan lambda_utils.LambdaFunctionType, MaxConcurrentDownload)
+
+	pushImageChannel := make(chan lambda_utils.LambdaFunctionType, MaxConcurrentUpload)
+
+	// Run authentication only once since the credentials will be recorded.
+	err = lambda_utils.AuthenticateDockerToECR()
+	if err != nil {
+		return errors.Wrap(err, "Unable to authenticate Lambda Function.")
+	}
+
+	// Pull images on a concurrency of "MaxConcurrentDownload".
+	errGroup.SetLimit(MaxConcurrentDownload)
+	go lambda_utils.AddFunctionTypeToChannel(functionsToShip[:], pullImageChannel)
+	for lambdaFunction := range pullImageChannel {
+		lambdaFunctionType := lambdaFunction
+		errGroup.Go(func() error {
+			return lambda_utils.PullImageFromECR(lambdaFunctionType)
+		})
```

Comment:
hmm this still doesn't seem quite like the right concurrency model. Let me know what you think about this:
- Typically in a situation where you have a maximum number of workers situation, you should spin up those workers explicitly, and have them constantly poll for new work once their old work is finished. Intuitively, this means that we should immediately spin up `MaxConcurrentDownload` workers with `errGroup.Go()`. Inside, each goroutine, you should fetch a function type from the FunctionType channel to start downloading. Then, that goroutine writes that processed FunctionType to a different, upload work channel, and goes back to polling on the input FunctionType channel until all work is completed.

- In the upload case, you would do the same thing, but poll on the output FunctionType channel from the above step. This allows a much more efficient pipeline where you can immediately start uploading pulled images to ECR, even while other pulls are taking place. Essentially, *all* the waiting should happen at the very end.

In order for all this to work, 1) `AddFunctionTypeToChannel` will have to happen synchronously in the very beginning. 2) You'll need one errgroup for uploads and one for downloads. 

Happy to chat about this! I agree it's kinda complicated, not super easy to test and iterate on, and might not be the highest priority thing over the other lambda tasks. Maybe we can try pairing on this sometime next week?

Reply:
Yeah works for me! I will experiment with the currency model you mention a bit before that.

Reply:
Cool feel free to drop something on my calendar :)